odoo/addons/base/transpiler/sourcemap_generator.py

```diff
@@ -0,0 +1,170 @@
+            "sources": list(self._sources.keys()),
+            "mappings": self._serialize_mappings(),
+            "sourcesContent": [self._sourcesContents[source] for source in self._sources]
+        }
+        if (self._file is not None):
+            mapping["file"] = self._file
+
+        if (self._sourceRoot is not None):
+            mapping["sourceRoot"] = self._sourceRoot
+
+        return mapping
+
+    def add_source(self, source_name, source_content, previous_genererated_line, nb_skip_line_output=0):
+        """Adds a new source file in the sourcemap. All the lines of the source file will be mapped line by line
+        to the generated file from the (previous_genererated_line + nb_skip_line_output). All lines between
+        previous_genererated_line and (previous_genererated_line + nb_skip_line_output) will
+        be mapped to line 1 of the source file.
+
+        Example:
+            ls 1 = Line 1 from new source file
+            lg 1 = Line 1 from genereted file
+            ls 1 <=> lg 1 Line 1 from new source file is map to  Line 1 from genereted file
+            nb_ls = number of lines in the new source file
+
+            Step 1:
+            ls 1 <=> lg previous_genererated_line + 1
+
+            Step 2:
+            ls 1 <=> lg previous_genererated_line + nb_skip_line_output + 1
+            ls 2 <=> lg previous_genererated_line + nb_skip_line_output + 2
+            ...
+            ls nb_ls <=> lg previous_genererated_line + nb_skip_line_output + nb_ls
+
+
+        :param source_name: name of the source to add
+        :param source_content: content of the source to add
+        :param previous_genererated_line: Line where we start to map the new source
+        :param nb_skip_line_output: Number of lines to pass in the generated file before starting mapping line by line
+
+        :return number of lines added
+        """
+        source_line_count = len(source_content.split("\n"))
+
+        if (source_name not in self._sources):
+            self._sources[source_name] = len(self._sources)
+
+        self._sourcesContents[source_name] = source_content
+        if nb_skip_line_output > 0:
+            # adds a mapping between the first line of the source
+            # and the first line of the corresponding code in the generated file.
+            self._mappings.append({
+                "generatedLine": previous_genererated_line + 1,
+                "originalLine": 1,
+                "source": source_name,
+            })
+        for i in range(1, source_line_count + 1):
+            self._mappings.append({
+                "generatedLine": previous_genererated_line + i + nb_skip_line_output,
+                "originalLine": i,
+                "source": source_name,
+            })
+        return source_line_count + nb_skip_line_output
+
+
+_b64chars = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
+shiftsize, flag, mask = 5, 1 << 5, (1 << 5) - 1
+
+
+@lru_cache(maxsize=None)
```

Comment:
No cache limit?  Please don't do this.

Reply:
What is the best size? (best practice) ?

Reply:
The best practice is "it depends" :wink: 
Try it on a realistic use-case, and see what size looks enough to make it effective.

Reply:
Probably need to run this on a few workloads and see how it behaves: basic stats can be obtained using `base64vlq_encode.cache_info()`. 

The uncertainty as to the workload and caching behaviour is why I didn't mind your original version (which had a cache for each instance of `SourceMapGenerator`, just didn't need to use an ad-hoc one) as that provided a well-defined lifetime for the resource.

Reply:
An other question is whether the matches are strongly temporal or not: if I understand correctly the VLQ encoding is done on the line numbers (original and generated), the question is whether the mappings are largely ordered with respect to those (and thus there'd be a high hit rate even with a small cache) or not.